### ISSUE TITLE
Enhance formatting of received array (or enumerable?) arguments in "unmatched call" output

### ISSUE THREAD

**Issue by TimLovellSmith** [FakeItEasy/FakeItEasy#1193]
I'm using FIE 2.3.0

I notice it's a little hard to figure out why my test failed with the current error messages. I am doing an assertion like this:

```
            A.CallTo(() => fake.CreateNewPortRules(
                A<IList<int>>.That.IsSameSequenceAs(new[] { 987, 123 }), 
                true))
            .MustHaveHappened();
```

And the output of the failed test is this:

```
Message: FakeItEasy.ExpectationException : 

  Assertion failed for the following call:
    RedisLaunch.PortManager.CreateNewPortRules(<specified sequence>, True)
  Expected to find it at least once but found it #0 times among the calls:
    1: RedisLaunch.PortManager.SSLPorts
    2: RedisLaunch.PortManager.CreateNewPortRules(ports: System.Int32[], allow: True)
    3: RedisLaunch.PortManager.ClearTextPorts
    4: RedisLaunch.PortManager.CreateNewPortRules(ports: System.Int32[], allow: True)
    5: RedisLaunch.PortManager.CreateStartupFirewallSettings(clearTextOpen: True, sslOpen: True)
```

**Comment by TimLovellSmith** [comment]
(Actually for my test I soon realized the problem is we get 2 calls with 1 item in the array instead of 1 call with 2.... but in general it might be a harder puzzle to crack without a debugger)

**Comment by blairconrad** [comment]
@TimLovellSmith, thanks for raising this.
The wrinkle, as I see it, is that there's currently no message "for `That.IsSameSequenceAs`". The list of received calls is independent of the expectations. There's no connection between a particular received call and an expected calls (if there were, we'd have a match! :wink:).

So I think the issue amounts to "enhance format of enumerable arguments in received calls". That sounds okay, but is tricky. One would have to account for very large enumerations and types that are enumerable but whose enumerableness is a secondary function (strings come to mind). And probably other things that I haven't thought of.

Or I suppose an incremental improvement would be to add a special format just for arrays. That doesn't alleviate my concern about very large arrays, but would avoid the "string problem".

Are you interested in working on this? Of course, we'd be targeting 4.x, not 2.3.y.

And don't forget, in the meantime, you can always add a custom [ArgumentValueFormatter](http://fakeiteasy.readthedocs.io/en/stable/formatting-argument-values/) as stopgap.

**Comment by blairconrad** [comment]
@thomaslevesque, @adamralph, I'm marking this as a potential enhancement. Suits?

**Comment by thomaslevesque** [comment]
> The wrinkle, as I see it, is that there's currently no message "for That.IsSameSequenceAs".

There *is* a message ("<specified sequence>"), but it was stripped from the original description because of the angled brackets ;). But it could certainly be more descriptive.

I guess we could output the values of the first few elements, and maybe the count if there are too many elements. Something like this:

```
  Assertion failed for the following call:
    RedisLaunch.PortManager.CreateNewPortRules(<987, 123>, True)
```
or
```
  Assertion failed for the following call:
    RedisLaunch.PortManager.CreateNewPortRules(<1, 2, 3, ... (10 more elements)>, True)
```
In addition, we could probably make the API a little friendlier by adding an overload with a `params` array.
And maybe a generic overload that works on `IEnumerable<T>` rather than `IEnumerable`.

**Comment by blairconrad** [comment]
> > The wrinkle, as I see it, is that there's currently no message "for That.IsSameSequenceAs".

> There *is* a message (`"<specified sequence>"`), but it was stripped from the original description because of the angled brackets ;). But it could certainly be more descriptive.

Sorry, I wasn't clear. I meant that the output of received calls isn't specialized based on the the unmatched expectation. So " 2: RedisLaunch.PortManager.CreateNewPortRules(ports: System.Int32[], allow: True)" looks the same regardless of the unmatched expectation.

I suppose we could break with that, applying knowledge of which argument failed to match a received call (if any) to produce a tailored rendering of the argument and the expectation. Seems ambitious.

Your idea to output some of the elements is a good one. It won't help in every case, but it would be hard to find something that would. And whatever rendering scheme is decided upon could be applied both to sequences in the expectation and in the received call.

> In addition, we could probably make the API a little friendlier by adding an overload with a `params` array.
> And maybe a generic overload that works on `IEnumerable<T>` rather than `IEnumerable`.

Absolutely, but I think in a different issue.

**Comment by thomaslevesque** [comment]
> Sorry, I wasn't clear. I meant that the output of received calls isn't specialized based on the the unmatched expectation. So " 2: RedisLaunch.PortManager.CreateNewPortRules(ports: System.Int32[], allow: True)" looks the same regardless of the unmatched expectation.

Ah, I see.

> I suppose we could break with that, applying knowledge of which argument failed to match a received call (if any) to produce a tailored rendering of the argument and the expectation. Seems ambitious.

Pretty ambitious, yeah...

> And whatever rendering scheme is decided upon could be applied both to sequences in the expectation and in the received call.

👍 

> Absolutely, but I think in a different issue.

OK, I'll open one ASAP

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.1.0).

Thanks for your help, @TimLovellSmith! Look for your name in the release notes. :trophy:
